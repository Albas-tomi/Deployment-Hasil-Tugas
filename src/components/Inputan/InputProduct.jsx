import { useFormik } from "formik";
import * as Yup from "yup";
import React from "react";
import { useDispatch } from "react-redux";
import { v4 as uuidv4 } from 'uuid';
import swal from "sweetalert";
import { useNavigate } from "react-router-dom";
import { createDataProduct } from "../productsThunk";
import { gql, useMutation } from "@apollo/client";

const Schema = Yup.object({
  name: Yup.string()
    .matches(/^[a-zA-Z ]+$/)
    .min(2, 'name must be at least 2 characters')
    .max(25)
    .required(),
  price: Yup.number().min(0).required(),
  category: Yup.string().required().matches(/^[a-zA-Z0-9_-]+/),
  image: Yup.mixed().required()
  // .test("FILE_SIZE","File Size to Big", (value) => value && value < 102400)
  .test("TYPE_FILE","Invalid",(value) => value && ['image/png', 'image/jpeg'].includes(value.type)),
  description: Yup.string()
  .matches(/^[a-zA-Z ]/).required(),
  freshness: Yup.string().required().matches(/^[a-zA-Z0-9_-]+/),
});

const RETRIEVE_PRODUCTS = gql`
query MyQuery {
    products {
      id
      name
      category
      description
      freshness
      price
    }
  }
`

const INSERT_PRODUCT_MUTATION =gql`
mutation AddNewProducts($name: String!, $category: String!, $freshness: String!, $description: String!, $price: Int!) {
  insert_products_one(object: {category: $category, description: $description, freshness: $freshness, name: $name, price: $price}) {
    id
  }
}
`

function InputProduct() {
  const [AddNewProducts, { data, loading , error}] = useMutation(
    INSERT_PRODUCT_MUTATION,
    {
      refetchQueries:[
        {
          query: RETRIEVE_PRODUCTS
      },
      'MyQuery'
      ]
    }
    );
  const navigate = useNavigate();
  // const dispatch = useDispatch();
  const formik = useFormik({
    initialValues: {
      id: "",
      name: "",
      category: "",
      image: '',
      price: "",
      description: "",
      freshness: "",
    },
    validationSchema: Schema,
    onSubmit: (values, { resetForm }) => {
      swal("Good job!", "You success input new product!", "success");
      AddNewProducts({
        variables:{
          name: values.name,
          category: values.category,
          price: values.price,
          description: values.description,
          freshness: values.freshness,
        }
      });
      navigate('/')
      resetForm();
    },
  });
  return (
    <div className=" bg-slate-200 mx-auto flex flex-col items-center p-10 gap-10 ">
      <form data-testid="form" name="form" onSubmit={formik.handleSubmit}>
        <div className="flex flex-col mb-3">
          <label className="text-xl" htmlFor="name">
            Name Product
          </label>
          <p data-testid="errors" className="text-red-500">{formik.errors.name}</p>
          <input
          className="rounded-md px-2"
            id="name"
            name="name"
            type="text"
            onChange={formik.handleChange}
            value={formik.values.name}
          />
        </div>
        <div className="mb-3 rounded-md">
          <label className="text-xl" htmlFor="category">
            Product Category
          </label>
          <p data-testid="errorCategory" className="text-red-500">{formik.errors.category}</p>
          <select
            className="flex flex-col"
            id="category"
            name="category"
            value={formik.values.category}
            onChange={formik.handleChange}
          >
            <option value={"0"}>
              Choose...
            </option>
            <option value={"Category 1"}>One</option>
            <option value={"Category 2"}>Two</option>
            <option value={"Category 3"}>Three</option>
          </select>
        </div>
        <div className="mb-3 flex flex-col">
          <label className="text-xl" htmlFor="image">
            Product Image
          </label>
          <p data-testid="errorImage" className="text-red-500">{formik.errors.image}</p>
          <input
           className="rounded-md"
            type="file"
            id="image"
            accept=".jpg,.jpeg,.png"
            name="image"
            // value={formik.values.image}
            onChange={(e) => formik.setFieldValue("image", e.target.files[0])}
          />
        </div>
          <label className="text-xl mb-3 " htmlFor="freshness">Freshness</label>
        <div>
        <p data-testid="errorFreshness" className="text-red-500">{formik.errors.freshness}</p>
          <input
           className="rounded-md m-3"
            id="freshness"
            name="freshness"
            type="radio"
            checked={"New brand" === formik.values.freshness}
            onChange={formik.handleChange}
            value={"New brand"}
          />
          <label htmlFor="freshness">New brand</label>
        </div>
        <div>
          <input
          className="m-3"
            id="name"
            name="freshness"
            type="radio"
            checked={"Second brand" === formik.values.freshness}
            onChange={formik.handleChange}
            value={"Second brand"}
          />
          <label htmlFor="freshness">Second brand</label>
        </div>
        <div>
          <input
          className="m-3"
            id="name"
            name="freshness"
            type="radio"
            checked={"Old brand" === formik.values.freshness}
            onChange={formik.handleChange}
            value={"Old brand"}
          />
          <label htmlFor="freshness">Old brand</label>
        </div>
  
        <div className="mb-3 flex flex-col">
          <label className="text-xl" htmlFor="description">
            Description Product
          </label>
          <p data-testid="errorDescription" className="text-red-500">{formik.errors.description}</p>
          <textarea
          rows={5}
           className="rounded-md p-2"
            id="description"
            name="description"
            type="number"
            onChange={formik.handleChange}
            value={formik.values.description}
          />
        </div>
        <div className="mb-3 flex flex-col">
          <label className="text-xl" htmlFor="price">
            Price Product
          </label>
          <p data-testid="errorPrice" className="text-red-500">{formik.errors.price}</p>
          <input
           className="rounded-md px-2"
            id="price"
            name="price"
            type="number"
            onChange={formik.handleChange}
            value={formik.values.price}
          />
        </div>
        <button
          className="bg-green-700 hover:bg-green-800 rounded-md p-2"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  );
}

export default InputProduct;
