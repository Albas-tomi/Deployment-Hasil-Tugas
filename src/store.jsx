import { configureStore } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";
import reducer from "./reducer";
import thunk from "redux-thunk";

const persistConfig = {
    key: "root",
    storage,
};

const  persistedReducer = persistReducer(persistConfig, reducer)

export const store = configureStore({
    reducer: persistedReducer,
    middleware: [thunk],
})

export const persiststore = persistStore(store);
