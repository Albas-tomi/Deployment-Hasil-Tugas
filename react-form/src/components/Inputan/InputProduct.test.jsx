import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
import InputProduct from "./InputProduct";
import { store } from "../../store";
import { Provider } from "react-redux";
import { vi } from "vitest";
import {fireEvent, render, screen, waitFor} from '@testing-library/react'

describe("InputProduct component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <InputProduct />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("submits the form correctly", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

//   memastikan bahwa form input Product Name dapat menerima input teks dan menampilkannya di halaman.
  it("name form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const input = getByLabelText("Name Product");
    act(() => {
      fireEvent.change(input, { target: { value: "Test Product" } });
    });
    expect(input.value).toBe("Test Product");
  });

  // Test agar name tidak kurang dari 2 abjad
  it("Name form input not <  2 character", async () => {
    const { getByLabelText, getByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const input = getByLabelText("Name Product");
    act(() => {
      fireEvent.change(input, { target: { value: "a" } });
      fireEvent.blur(input);
    });
    await waitFor(() => {
      expect(getByTestId("errors")).toBeInTheDocument();
    });
    expect(getByTestId("errors")).toHaveTextContent("name must be at least 2 characters");

  });

  // Test agar name tidak lebih dari 25 abjad
  it("Name form input not > 25 character", async () => {
    const { getByLabelText, getByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const input = getByLabelText("Name Product");
    act(() => {
      fireEvent.change(input, { target: { value: "aaaaaaaaaaaaaaaaaaaaaaaaaa" } });
      fireEvent.blur(input);
    });
    await waitFor(() => {
      expect(getByTestId("errors")).toBeInTheDocument();
    });
    expect(getByTestId("errors")).toHaveTextContent("name must be at most 25 characters");

  });
  // Test agar name tidak mengandung @/#{}
  it("Name form input not contains @/#{} ", async () => {
    const { getByLabelText, getByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const input = getByLabelText("Name Product");
    act(() => {
      fireEvent.change(input, { target: { value: "@/#{}" } });
      fireEvent.blur(input);
    });
    await waitFor(() => {
      expect(getByTestId("errors")).toBeInTheDocument();
    });
    expect(getByTestId("errors")).toHaveTextContent(`name must match the following: "/^[a-zA-Z ]+$/"`);
  });


//   memastikan bahwa form input Product Category dapat menerima input teks dan menampilkannya di halaman.
  it("Category  form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const inputCategory = getByLabelText("Product Category");
    act(() => {
      fireEvent.change(inputCategory, { target: { value: 'Category 1' } });
    });
    expect(inputCategory.value).toBe('Category 1');
  });

 

  //   memastikan bahwa form input Product Image dapat menerima dan menampilkannya di halaman.
  it("Image form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const inputImage = getByLabelText("Product Image");
    act(() => {
      fireEvent.input(inputImage, { target: { value: '' } });
      fireEvent.change(inputImage, { target: { files: [new File(['image.png'], 'image.png', { type: 'image/png' })] } });
    });
    expect(inputImage.files[0].name).toBe('image.png');
  });

//   memastikan bahwa form input Product Freshness dapat menerima input teks dan menampilkannya di halaman.
  it("Freshness form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const inputFreshness = getByLabelText("Freshness");
    act(() => {
      fireEvent.change(inputFreshness, { target: { value: 'hayyy' } });
    });
    expect(inputFreshness.value).toBe('hayyy');
  });

//   memastikan bahwa form input Product Description dapat menerima input teks dan menampilkannya di halaman.
  it("Description form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const inputDescription = getByLabelText("Description Product");
    act(() => {
      fireEvent.change(inputDescription, { target: { value: 'hayyy' } });
    });
    expect(inputDescription.value).toBe('hayyy');
  });

//   memastikan bahwa form input Product Price dapat menerima input teks dan menampilkannya di halaman.
  it("price form input correctly", () => {
    const { getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={vi.fn()} />
        </MemoryRouter>
      </Provider>
    );
    const inputPrice = getByLabelText("Price Product");
    act(() => {
      fireEvent.change(inputPrice, { target: { value: '2323' } });
    });
    expect(inputPrice.value).toBe('2323');
  });


//  Testing Semua Field Tidak boleh kosong
  it("form submit input correctly", async () => {
    //mocksubmit untuk fake user submit
    const mockSubmit = vi.fn()
    const { getByTestId, getByLabelText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <InputProduct onSubmit={mockSubmit} />
        </MemoryRouter>
      </Provider>
    ); 
    const form = getByTestId("form");
    const input = getByLabelText("Name Product");
    const inputCategory = getByLabelText("Product Category");
    const inputImage = getByLabelText("Product Image");
    const inputFreshness = getByLabelText("Freshness");
    const inputDescription = getByLabelText("Description Product");
    const inputPrice = getByLabelText("Price Product");
    act(() => {
      // user click submit
      fireEvent.change(input, { target: { value: "" } });
      fireEvent.change(inputCategory, { target: { value: '' } });
      fireEvent.input(inputImage, { target: { value: '' } });
      fireEvent.change(inputFreshness, { target: { value: '' } });
      fireEvent.change(inputDescription, { target: { value: '' } });
      fireEvent.change(inputPrice, { target: { value: '' } });
      fireEvent.submit(form);
    });
    await waitFor(() => {
      expect(getByTestId("errors")).toHaveTextContent(`name is a required field`);
      expect(getByTestId("errorCategory")).toHaveTextContent(`category is a required field`);
      expect(getByTestId("errorImage")).toHaveTextContent(`image is a required field`);
      expect(getByTestId("errorFreshness")).toHaveTextContent(`freshness is a required field`);
      expect(getByTestId("errorDescription")).toHaveTextContent(`description is a required field`);
      expect(getByTestId("errorPrice")).toHaveTextContent(`price is a required field`);

    });
  });
});
