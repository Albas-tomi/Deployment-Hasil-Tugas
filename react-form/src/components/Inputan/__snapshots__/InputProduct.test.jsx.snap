// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputProduct component > renders correctly 1`] = `
<div
  className=" bg-slate-200 mx-auto flex flex-col items-center p-10 gap-10 "
>
  <form
    data-testid="form"
    name="form"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col mb-3"
    >
      <label
        className="text-xl"
        htmlFor="name"
      >
        Name Product
      </label>
      <p
        className="text-red-500"
        data-testid="errors"
      />
      <input
        className="rounded-md px-2"
        id="name"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3 rounded-md"
    >
      <label
        className="text-xl"
        htmlFor="category"
      >
        Product Category
      </label>
      <p
        className="text-red-500"
        data-testid="errorCategory"
      />
      <select
        className="flex flex-col"
        id="category"
        name="category"
        onChange={[Function]}
        value=""
      >
        <option
          value="0"
        >
          Choose...
        </option>
        <option
          value="Category 1"
        >
          One
        </option>
        <option
          value="Category 2"
        >
          Two
        </option>
        <option
          value="Category 3"
        >
          Three
        </option>
      </select>
    </div>
    <div
      className="mb-3 flex flex-col"
    >
      <label
        className="text-xl"
        htmlFor="image"
      >
        Product Image
      </label>
      <p
        className="text-red-500"
        data-testid="errorImage"
      />
      <input
        accept=".jpg,.jpeg,.png"
        className="rounded-md"
        id="image"
        name="image"
        onChange={[Function]}
        type="file"
      />
    </div>
    <label
      className="text-xl mb-3 "
      htmlFor="freshness"
    >
      Freshness
    </label>
    <div>
      <p
        className="text-red-500"
        data-testid="errorFreshness"
      />
      <input
        checked={false}
        className="rounded-md m-3"
        id="freshness"
        name="freshness"
        onChange={[Function]}
        type="radio"
        value="New brand"
      />
      <label
        htmlFor="freshness"
      >
        New brand
      </label>
    </div>
    <div>
      <input
        checked={false}
        className="m-3"
        id="name"
        name="freshness"
        onChange={[Function]}
        type="radio"
        value="Second brand"
      />
      <label
        htmlFor="freshness"
      >
        Second brand
      </label>
    </div>
    <div>
      <input
        checked={false}
        className="m-3"
        id="name"
        name="freshness"
        onChange={[Function]}
        type="radio"
        value="Old brand"
      />
      <label
        htmlFor="freshness"
      >
        Old brand
      </label>
    </div>
    <div
      className="mb-3 flex flex-col"
    >
      <label
        className="text-xl"
        htmlFor="description"
      >
        Description Product
      </label>
      <p
        className="text-red-500"
        data-testid="errorDescription"
      />
      <textarea
        className="rounded-md p-2"
        id="description"
        name="description"
        onChange={[Function]}
        rows={5}
        type="number"
        value=""
      />
    </div>
    <div
      className="mb-3 flex flex-col"
    >
      <label
        className="text-xl"
        htmlFor="price"
      >
        Price Product
      </label>
      <p
        className="text-red-500"
        data-testid="errorPrice"
      />
      <input
        className="rounded-md px-2"
        id="price"
        name="price"
        onChange={[Function]}
        type="number"
        value=""
      />
    </div>
    <button
      className="bg-green-700 hover:bg-green-800 rounded-md p-2"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
